src/Service/MemoryInfoHandler.h

```diff
@@ -36,10 +38,18 @@ class MemoryInfoHandler : public orbit_memory_tracing::MemoryInfoListener {
   void OnProcessMemoryUsage(orbit_grpc_protos::ProcessMemoryUsage process_memory_usage) override;
 
  private:
+  [[nodiscard]] uint64_t GetSamplingWindowId(uint64_t sample_timestamp) const;
+  void ProcessMemoryEventWrapperIfReady(uint64_t sampling_window_id);
+
   ProducerEventProcessor* producer_event_processor_;
   std::unique_ptr<orbit_memory_tracing::MemoryInfoProducer> cgroup_memory_info_producer_;
   std::unique_ptr<orbit_memory_tracing::MemoryInfoProducer> process_memory_info_producer_;
   std::unique_ptr<orbit_memory_tracing::MemoryInfoProducer> system_memory_info_producer_;
+  uint64_t sampling_start_timestamp_ns_;
+  uint64_t sampling_period_ns_;
+  bool enable_cgroup_memory_ = false;
+  absl::flat_hash_map<uint64_t, orbit_grpc_protos::MemoryEventWrapper> in_progress_wrappers_;
```

Comment:
Nit: You could use abseil's `GUARDED_BY` to make clear that this map is synchronized by the mutex below. See https://abseil.io/docs/cpp/guides/synchronization#annotation-guidelines for details.

Reply:
Changed accordingly. Now this part is moved in to `MemoryInfoListener.h`. Thanks for the suggestion!